### PR TITLE
Main parquet text loader

### PR DESCRIPTION
**Pull request by artemru**
**What does this PR do? Please describe:**
* Enable text dataloading from parquet dataset (shared at row groups levels)
* adding split to text and jsonl dataloader reader to unify the interface
* fixing npc at other different dataloader

**Does your PR introduce any breaking changes? If yes, please list them:**
No BC, only new feature

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [ ] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/datasets/jsonl.py [hunk 1]
```diff
@@ -110,28 +144,26 @@ def encode(example: dict[str, Any]) -> Tensor:
             max_num_elements = batching.max_num_elements
 
             builder.pack(
-                max_num_elements + 1, max_seq_len, truncate=True, pinned_memory=True
+                max_num_elements + 1,
+                max_seq_len,
+                pad_value=pad_idx,
```
Comment (artemru) [inline src/fairseq2/datasets/jsonl.py:158]: @cbalioglu : pad_value is not documented, I hope I used it correctly

File: src/fairseq2/datasets/text.py [hunk 2]
```diff
@@ -123,16 +149,20 @@ def create_reader(
         min_seq_len: int,
         max_seq_len: int,
         options: TextReadOptions | None = None,
+        split: str | None = None,
```
Comment (artemru) [inline src/fairseq2/datasets/text.py:165]: this will keep the previous behaviour by default

File: src/fairseq2/datasets/text_parquet.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        if split is not None:
+            fragment_config = fragment_config.add_partition_filter(
+                pa.compute.field("split") == split
+            )
+        fragement_builder = ParquetFragmentStreamer(
+            config=fragment_config
+        ).build_pipeline(rank=rank, world_size=world_size)
+
+        num_parallel_fragments = options.extras.get("num_parallel_fragments", npc)
+        assert isinstance(num_parallel_fragments, int)
```
Comment (zyaoj) [inline src/fairseq2/datasets/text_parquet.py]: For the `assert`s, could we add more error messages to make it as instructive as possible?

File: src/fairseq2/datasets/text_parquet.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        )
+        options.seed += gang.rank
+
+        pipeline = JsonlDataset.build_pipeline_backend(
+            builder,
+            options,
+            text_encoder,
+            pad_idx=pad_idx,
+            max_seq_len=max_seq_len,
+            text_column_name="text",
```
Comment (zyaoj) [inline src/fairseq2/datasets/text_parquet.py:225]: This hardcode the text column name to "text" instead of allowing dynamic configuration of the target. For example say if I want to load `sampled_sonar_wiki_lingua.parquet`, whose `source` and `target` columns are text columns.  Shall we set `text_column_name` as an input argument to this functional interface?

File: src/fairseq2/datasets/instruction.py [hunk 5]
```diff
@@ -238,8 +234,8 @@ def create_reader(
         source_encoder = tokenizer.create_encoder(mode=options.source_encode_mode)
         target_encoder = tokenizer.create_encoder(mode=options.target_encode_mode)
 
-        builder.map(source_encoder, selector="src", num_parallel_calls=npc)
-        builder.map(target_encoder, selector="tgt", num_parallel_calls=npc)
+        builder.map(source_encoder, selector="src", num_parallel_calls=1)
```
Comment (zyaoj) [inline src/fairseq2/datasets/instruction.py]: shall we change this implicit default to an `options` field for early-stage/light work?

File: src/fairseq2/datasets/text.py [hunk 6]
```diff
@@ -114,6 +115,31 @@ def from_path(path: Path, name: str) -> GenericTextDataset:
...
+        """Filter the dataset files by split.
+
+        :param split:
+            The split to filter by.
+        """
+        if split is None:
+            return files
+
+        # filtering based on two following patterns:
+        pattern1 = f"**/*{split}*.{extention}"
```
Comment (zyaoj) [inline src/fairseq2/datasets/text.py]: is this going to fail for some cases like `constraint.parquet`?

File: src/fairseq2/datasets/text_parquet.py (hunk 4 above)
Comment (zyaoj) [reply src/fairseq2/datasets/text_parquet.py:225]: After hardcoding the column to `source` I still got `KeyError: 'source'`. Is there something I'm missing?

File: src/fairseq2/datasets/text.py (hunk 6 above)
Comment (artemru) [reply src/fairseq2/datasets/text.py]: this's only for jsonl or text examples

File: src/fairseq2/datasets/text_parquet.py (hunk 4 above)
Comment (artemru) [reply src/fairseq2/datasets/text_parquet.py:225]: you need to change the value of `DefaultTextSchema.text = "source" ! note that this class is universally responsible of renaming input columns into the target schema.
That being said we can use only add an extra param here.

Comment (artemru) [reply src/fairseq2/datasets/text_parquet.py:225]: see the doc string for DefaultTextSchema

File: src/fairseq2/datasets/instruction.py (hunk 5 above)
Comment (artemru) [reply src/fairseq2/datasets/instruction.py]: I dont completely get the idea, but here in any case, putting num_parallel_calls>1 will only result in the slowdown since the fn is GIL strongly bound.

Comment (zyaoj) [reply src/fairseq2/datasets/instruction.py]: I think the idea overall is to avoid magic numbers. There could be 2 options:
- add some comments there to note the choice of this config
- set it as a field in the options as something like `npc_gil_bound_job`

Comment (artemru) [reply src/fairseq2/datasets/instruction.py]: ok, we can rm num_parallel_calls=1 here at all as an arg to avoid an extra param at all

File: src/fairseq2/datasets/text.py (hunk 6 above)
Comment (zyaoj) [reply src/fairseq2/datasets/text.py]: Yeah I think the point here is not about the file extension, but more about the regex pattern for file names. This might be a nit but there could be some edge cases like:
- when `split = "train"`, if the file is named as `wiki-constraint-test.jsonl` or `wiki-constraint-test.*`, though it's probably a test split dataset, it contains `train` in `constraint` which might accidentally include a test split file into the train split. Wdyt?

File: src/fairseq2/datasets/text_parquet.py (hunk 4 above)
Comment (zyaoj) [reply src/fairseq2/datasets/text_parquet.py:225]: Ahah I indeed missed the `DefaultTextSchema` and this works like a charm.

File: src/fairseq2/datasets/text.py (hunk 6 above)
Comment (artemru) [reply src/fairseq2/datasets/text.py]: ah, I see, ok what about the strict check ? so only train.jsonl will be accepted ?

Comment (artemru) [reply src/fairseq2/datasets/text.py]: ( I went for most the strict schema), somehow for text files there's no universal conversion for partitioning unlikely for parquet)

Comment (zyaoj) [reply src/fairseq2/datasets/text.py]: Yeah I guess we can put a strict check as a default and add an input argument that accepts split patterns, something like:

```py
@staticmethod
def filter_split(
    files: Sequence[Path],
    split: str | None,
    extention: str,
    *,
    split_pattern: str | Sequence[str] | None = None,  # ← change here
) -> Sequence[Path]:
    if split is None:
        return list(files)

    # Prepare pattern list
    if split_pattern is None:
        patterns: list[str] = [
            f"**/{split}.{extention}",
            f"{split}/*.{extention}",
        ]
    else:
        # Accept str, list[str], tuple[str], … and interpolate placeholders
        raw = [split_pattern] if isinstance(split_pattern, str) else list(split_pattern)
        patterns = [pattern.format(split=split, ext=extention) for pattern in raw]

    # Filter paths 
    return [path for path in files if any(fnmatch.fnmatch(str(path), pattern) for pattern in patterns)]
```

Comment (artemru) [reply src/fairseq2/datasets/text.py]: good idea indeed !

Comment (artemru) [reply src/fairseq2/datasets/text.py]: I took it

**Review by zyaoj — APPROVED** [review]
LGTM! Thanks for addressing the comments.
